tests/functional/olp-cpp-sdk-dataservice-read/DataserviceReadVolatileLayerClientTest.cpp

```diff
@@ -224,4 +341,273 @@ TEST_F(VolatileLayerClientTest, GetPartitionsDifferentFetchOptions) {
+
+  PrepareForPrefetch();
+
+  olp::client::HRN hrn(catalog);
+  VolatileLayerClient client(hrn, layer, settings_);
+  {
+    SCOPED_TRACE("Prefetch tiles online and store them in memory cache");
+    std::vector<olp::geo::TileKey> tile_keys = {
+        olp::geo::TileKey::FromHereTile(kTileId)};
+
+    auto request = olp::dataservice::read::PrefetchTilesRequest()
+                       .WithTileKeys(tile_keys)
+                       .WithMinLevel(10)
+                       .WithMaxLevel(12);
+
+    std::promise<PrefetchTilesResponse> promise;
+    std::future<PrefetchTilesResponse> future = promise.get_future();
+    auto token = client.PrefetchTiles(
+        request, [&promise](PrefetchTilesResponse response) {
+          promise.set_value(std::move(response));
+        });
+
+    ASSERT_NE(future.wait_for(kTimeout), std::future_status::timeout);
+    PrefetchTilesResponse response = future.get();
+    EXPECT_TRUE(response.IsSuccessful());
+    ASSERT_FALSE(response.GetResult().empty());
+
+    const auto& result = response.GetResult();
+
+    for (auto tile_result : result) {
+      EXPECT_TRUE(tile_result->IsSuccessful());
+      ASSERT_TRUE(tile_result->tile_key_.IsValid());
+    }
+
+    ASSERT_EQ(3u, result.size());
+  }
+
+  {
+    SCOPED_TRACE("Read cached data from the same partition");
+    DataResponse response;
+    auto token = client.GetData(
+        olp::dataservice::read::DataRequest()
+            .WithPartitionId(kTileId)
+            .WithFetchOption(CacheOnly),
+        [&response](DataResponse resp) { response = std::move(resp); });
+    EXPECT_TRUE(response.IsSuccessful());
+    ASSERT_TRUE(response.GetResult() != nullptr);
+    ASSERT_NE(response.GetResult()->size(), 0u);
+  }
+
+  {
+    SCOPED_TRACE("Read cached data from pre-fetched sub-partition #1");
+    const auto sub_partition1 = CustomParameters::getArgument(kSubPartitionId1);
+    DataResponse response;
+    auto token = client.GetData(
+        olp::dataservice::read::DataRequest()
+            .WithPartitionId(sub_partition1)
+            .WithFetchOption(CacheOnly),
+        [&response](DataResponse resp) { response = std::move(resp); });
+    EXPECT_TRUE(response.IsSuccessful());
+    ASSERT_TRUE(response.GetResult() != nullptr);
+    ASSERT_NE(response.GetResult()->size(), 0u);
+  }
+
+  {
+    SCOPED_TRACE("Read cached data from pre-fetched sub-partition #2");
+    auto sub_partition2 = CustomParameters::getArgument(kSubPartitionId2);
+    DataResponse response;
+    auto token = client.GetData(
+        olp::dataservice::read::DataRequest()
+            .WithPartitionId(sub_partition2)
+            .WithFetchOption(CacheOnly),
+        [&response](DataResponse resp) { response = std::move(resp); });
+    EXPECT_TRUE(response.IsSuccessful());
+    ASSERT_TRUE(response.GetResult() != nullptr);
+    ASSERT_NE(response.GetResult()->size(), 0u);
+  }
+}
+
+TEST_F(VolatileLayerClientTest, PrefetchWrongLevels) {
```

Comment:
These are all integration tests.
The only test you need to do extra is that in case min/max are equal or 0, that only the requested tilekeys are downloaded.

Reply:
Removed.
Extra tests for 0/0 and same levels added.